README.md
## Flask-Passlib ##

Flask extension for [Passlib](http://packages.python.org/passlib/).

This is still WIP, but give it a try.

### Installation ###

```pip install Flask-Passlib```

### Usage ###

The default usage will attempt to figure out what version of ```werkzeug``` is being used. 
As of ```werkzeug <= 0.8.3``` only supports MD5 and SHA1. 
```python
app = Flask(__name__)
passlib = Passlib(app)
```

To change the ```passlib``` schemes, initialize with an instance of CryptContext and set the ```default``` scheme.

```python
from passlib.context import LaxyCryptContext
from flask.ext.passlib.context import (werkzeug_salted_md5, werkzeug_salted_sha1, 
    werkzeug_salted_sha256, werkzeug_salted_sha512)
from passlib.hash import django_pbkdf2_sha256

passlib = Passlib(app, context=LazyCryptContext(
    schemes=[
        werkzeug_salted_md5,
        werkzeug_salted_sha1,
        werkzeug_salted_sha256,
        werkzeug_salted_sha512,         

        django_pbkdf2_sha256,
    ],
    default='django_pbkdf2_sha256',
))
```

In this example, we have the Flask defaults with __Django's PBKDF2 SHA256__.

##### Verifying #####

In Flask using werkzeug's

```python
check_password_hash(password_hash, raw_password)
```

With Flask-Passlib

```python
passlib.verify(raw_password, password_hash)
```

##### Generating #####

In Flask using werkzeug's

```python
generate_password_hash(raw_password, method=.., salt_length=..)
```

With Flask-Passlib

```python
passlib.encrypt(raw_password, method=..., salt_length=...)
```
### Documentation ###
Also check out [Passlib's documentation](http://packages.python.org/passlib/)


CHANGES
Flask-Passlib Changelog
====================


Version 0.1
---------------------

Released Jan, 11 2013

 - Initial development

flask_passlib/tests.py
# -*- coding: utf-8 -*-
import unittest
import random
import string

from flask import Flask
from flask.ext.passlib import Passlib

from werkzeug.security import generate_password_hash, check_password_hash

WERKZEUG_METHODS = ['md5', 'sha1']
PASSLIB_METHODS = ['werkzeug_salted_md5', 'werkzeug_salted_sha1']

class FlaskTestCase(unittest.TestCase):
    TESTING = True

    def setUp(self):
        self.app = Flask(__name__)
        self.app.config.from_object(self)

        self.assertTrue(self.app.testing)

        self.passlib = Passlib(self.app)

        self.ctx = self.app.test_request_context()
        self.ctx.push()

    def tearDown(self):
        self.ctx.pop()

    def generate_password(self, length=6):
        return ''.join(random.choice(string.ascii_uppercase + string.digits) for x in range(length))

class TestCase(object):
    ident = None
    werkzeug_method = None
    passlib_method = None
    salt_length = 8
    checksum_length = 0
    seperator = '$'

    def test_identify(self):
        password = self.generate_password()

        # Identify hashes generated by werkzeug
        passlib_hash = generate_password_hash(password, method=self.werkzeug_method)
        self.assertEqual(self.passlib.identify(passlib_hash), self.passlib_method)

        # Identify hashes generated by other werkzeug methods
        diff_werkzeug_method = random.choice([method for method in WERKZEUG_METHODS if method not in [self.werkzeug_method]])
        passlib_hash = generate_password_hash(password, method=diff_werkzeug_method)
        self.assertNotEqual(self.passlib.identify(passlib_hash), self.passlib_method)

        # Identify hashes generated by werkzeug with varying salt length
        passlib_hash = generate_password_hash(password, method=self.werkzeug_method, salt_length=self.salt_length + 1)
        self.assertEqual(self.passlib.identify(passlib_hash), self.passlib_method)

        # Identify hashes generated by itself
        passlib_hash = self.passlib.encrypt(password, method=self.passlib_method)
        self.assertEqual(self.passlib.identify(passlib_hash), self.passlib_method)

        # Identify hashes generated by other methods
        diff_passlib_method = random.choice([method for method in PASSLIB_METHODS if method not in [self.passlib_method]])
        passlib_hash = self.passlib.encrypt(password, method=diff_passlib_method)
        self.assertNotEqual(self.passlib.identify(passlib_hash), self.passlib_method)

        # Identify hashes generated by itself with varying salt length
        passlib_hash = self.passlib.encrypt(password, method=self.passlib_method, salt_length=self.salt_length + 1)
        self.assertEqual(self.passlib.identify(passlib_hash), self.passlib_method)

    def test_verify(self):
        password = self.generate_password()
        werkzeug_hash = generate_password_hash(password, method=self.werkzeug_method)
        passlib_hash = self.passlib.encrypt(password, scheme=self.passlib_method)

        self.assertTrue(self.passlib.verify(password, werkzeug_hash))
        self.assertTrue(self.passlib.verify(password, passlib_hash))


        password2 = self.generate_password()
        werkzeug_hash2 = generate_password_hash(password2, method=self.werkzeug_method)
        passlib_hash2 = self.passlib.encrypt(password2, scheme=self.passlib_method)

        self.assertTrue(self.passlib.verify(password2, werkzeug_hash2))
        self.assertTrue(self.passlib.verify(password2, passlib_hash2))

        self.assertFalse(self.passlib.verify(password, werkzeug_hash2))
        self.assertFalse(self.passlib.verify(password, passlib_hash2))
        self.assertFalse(self.passlib.verify(password2, werkzeug_hash))
        self.assertFalse(self.passlib.verify(password2, passlib_hash))

    def test_encrypt(self):
        password = self.generate_password()

        passlib_hash = self.passlib.encrypt(password, scheme=self.passlib_method, salt_length=self.salt_length)
        self.assertTrue(check_password_hash(passlib_hash, password))
        self.assertIn(self.ident, passlib_hash)
        self.assertTrue(len(passlib_hash.split(self.seperator)), 3)
        self.assertTrue(passlib_hash.startswith(self.ident))
        self.assertTrue(len(passlib_hash.split(self.seperator)[1]), self.salt_length)
        self.assertTrue(len(passlib_hash.split(self.seperator)[2]), self.checksum_length)


class MD5TestCase(TestCase, FlaskTestCase):
    ident = 'md5$'
    werkzeug_method = 'md5'
    passlib_method = 'werkzeug_salted_md5'
    checksum_length = 32

class SHA1TestCase(TestCase, FlaskTestCase):
    ident = 'sha1$'
    werkzeug_method = 'sha1'
    passlib_method = 'werkzeug_salted_sha1'
    checksum_length = 40

class SHA224TestCase(TestCase, FlaskTestCase):
    ident = 'sha224$'
    werkzeug_method = 'sha224'
    passlib_method = 'werkzeug_salted_sha224'
    checksum_length = 56

class SHA256TestCase(TestCase, FlaskTestCase):
    ident = 'sha256$'
    werkzeug_method = 'sha256'
    passlib_method = 'werkzeug_salted_sha256'
    checksum_length = 64

class SHA384TestCase(TestCase, FlaskTestCase):
    ident = 'sha384$'
    werkzeug_method = 'sha384'
    passlib_method = 'werkzeug_salted_sha384'
    checksum_length = 96

class SHA512TestCase(TestCase, FlaskTestCase):
    ident = 'sha512$'
    werkzeug_method = 'sha512'
    passlib_method = 'werkzeug_salted_sha512'
    checksum_length = 128


            
if __name__ == '__main__':
    unittest.main()

flask_passlib/context.py
from distutils.version import StrictVersion
from passlib.context import LazyCryptContext
import werkzeug

from .handlers import (werkzeug_salted_md5, werkzeug_salted_sha1, 
    werkzeug_salted_sha224, werkzeug_salted_sha256, 
    werkzeug_salted_sha384, werkzeug_salted_sha512)


werkzeug061_context = LazyCryptContext(
    schemes=[
        werkzeug_salted_md5,    
        werkzeug_salted_sha1,
    ],
    default='werkzeug_salted_sha1',
)

werkzeugdev_context = LazyCryptContext(
    schemes=[
        werkzeug_salted_md5,    
        werkzeug_salted_sha1,
        werkzeug_salted_sha224,
        werkzeug_salted_sha256,
        werkzeug_salted_sha384,
        werkzeug_salted_sha512,
    ],
    default='werkzeug_salted_sha1',
)


if StrictVersion('0.6.1') <= StrictVersion(werkzeug.__version__) <= StrictVersion('0.8.3'):
    werkzeug_context = werkzeug061_context
elif StrictVersion(werkzeug.__version__) > StrictVersion('0.8.3'):
    werkzeug_context = werkzeugdev_context
else:
    werkzeug_context = werkzeug061_context

flask_context = werkzeug_context

flask_passlib/__init__.py
from .context import flask_context
from passlib.context import LazyCryptContext


class Passlib(object):
    def __init__(self, app=None, **kwargs):
        if app:
            self.init_app(app, **kwargs)

    def init_app(self, app, context=flask_context, schemes=None):
        self.app = app
        
        # register extension with app
        self.app.extensions = getattr(app, 'extensions', {})
        self.app.extensions['passlib'] = self

        self.context = context

    def verify(self, *args, **kwargs):
        return self.context.verify(*args, **kwargs)

    def encrypt(self, *args, **kwargs):
        if 'salt_length' in kwargs:
            kwargs['salt_size'] = kwargs.pop('salt_length')
        if 'method' in kwargs:
            kwargs['scheme'] = kwargs.pop('method')
        return self.context.encrypt(*args, **kwargs)

    def identify(self, *args, **kwargs):
        return self.context.identify(*args, **kwargs)

flask_passlib/handlers.py
from werkzeug.security import SALT_CHARS

from passlib.utils.compat import u, str_to_uascii
from passlib.utils import classproperty
import passlib.utils.handlers as uh

import hmac
from hashlib import md5, sha1, sha224, sha256, sha384, sha512


class WerkzeugGenericHandler(uh.GenericHandler):
    checksum_chars = uh.LOWER_HEX_CHARS
    werkzeug_hash_method = None

    # @classmethod
    # def identify(cls, hash):
    #     if hash.count('$') < 2:
    #         return False
    #     method, salt, checksum = hash.split('$', 2)

    #     # does class specify a known unique prefix to look for?
    #     ident = cls.ident
    #     if ident is not None:
    #         return (method + '$').startswith(ident)
    #     return False

    @classproperty
    def _stub_checksum(cls):
        return cls.checksum_chars[0] * cls.checksum_size

class WerkzeugSaltedHash(uh.HasSalt):
    """base class providing common code for Werkzeug hashes"""
    # name, ident, checksum_size must be set by subclass.
    # ident must include "$" suffix.
    setting_kwds = ("salt", "salt_size")

    min_salt_size = 0
    default_salt_size = 8
    max_salt_size = None
    salt_chars = SALT_CHARS

    def _calc_checksum(self, secret, digestmod=None):
        if isinstance(secret, unicode):
            secret = secret.encode('utf-8')
        if isinstance(self.salt, unicode):
            self.salt = self.salt.encode('utf-8')
        # return hmac.new(self.salt, secret, digestmod).hexdigest()
        return str_to_uascii(hmac.new(self.salt, secret, digestmod).hexdigest())

    @classmethod
    def from_string(cls, hash):
        salt, chk = uh.parse_mc2(hash, cls.ident, handler=cls)
        return cls(salt=salt, checksum=chk)

    def to_string(self):
        return uh.render_mc2(self.ident, self.salt,
                             self.checksum or self._stub_checksum)




class werkzeug_salted_md5(WerkzeugSaltedHash, WerkzeugGenericHandler):
    name = "werkzeug_salted_md5"
    ident = u("md5$")
    checksum_size = 32

    def _calc_checksum(self, secret, digestmod=md5):
        return super(werkzeug_salted_md5, self)._calc_checksum(secret, digestmod=digestmod)

class werkzeug_salted_sha1(WerkzeugSaltedHash, WerkzeugGenericHandler):
    name = "werkzeug_salted_sha1"
    ident = u("sha1$")
    checksum_size = 40

    def _calc_checksum(self, secret, digestmod=sha1):
        return super(werkzeug_salted_sha1, self)._calc_checksum(secret, digestmod=digestmod)

class werkzeug_salted_sha224(WerkzeugSaltedHash, WerkzeugGenericHandler):
    name = "werkzeug_salted_sha224"
    ident = u("sha224$")
    checksum_size = 56

    def _calc_checksum(self, secret, digestmod=sha224):
        return super(werkzeug_salted_sha224, self)._calc_checksum(secret, digestmod=digestmod)

class werkzeug_salted_sha256(WerkzeugSaltedHash, WerkzeugGenericHandler):
    name = "werkzeug_salted_sha256"
    ident = u("sha256$")
    checksum_size = 64

    def _calc_checksum(self, secret, digestmod=sha256):
        return super(werkzeug_salted_sha256, self)._calc_checksum(secret, digestmod=digestmod)

class werkzeug_salted_sha384(WerkzeugSaltedHash, WerkzeugGenericHandler):
    name = "werkzeug_salted_sha384"
    ident = u("sha384$")
    checksum_size = 96

    def _calc_checksum(self, secret, digestmod=sha384):
        return super(werkzeug_salted_sha384, self)._calc_checksum(secret, digestmod=digestmod)

class werkzeug_salted_sha512(WerkzeugSaltedHash, WerkzeugGenericHandler):
    name = "werkzeug_salted_sha512"
    ident = u("sha512$")
    checksum_size = 128

    def _calc_checksum(self, secret, digestmod=sha512):
        return super(werkzeug_salted_sha512, self)._calc_checksum(secret, digestmod=digestmod)